### PR TITLE
Backport of Onboard consul to use new version/VERSION file and reproducible actions into release/1.14.x

### PR DESCRIPTION
**Pull request by hc-github-team-consul-core**
## Backport

This PR is auto-generated from #15631 to be assessed for backporting due to the inclusion of the label backport/1.14.

**WARNING** automatic cherry-pick of commits failed. Commits will require human attention.

> merge conflict error: POST https://api.github.com/repos/hashicorp/consul/merges: 409 Merge conflict []

The below text is copied from the body of the original PR.

---

# PR Description
This PR introduces the new [actions-set-product-version](https://github.com/hashicorp/actions-set-product-version), a tiny, but mighty, github action that acts as a bridge between the product repo and our new CRT feature: [automated version bumping](https://github.com/hashicorp/bob/commit/6813d9757c644679193a0af317e99570ac8cc848). This PR also introduces automatic version patch bumping via the bump-version {todo: enter link when merged} workflow.

tl;dr automated version bumping has a new command (bob update version) in the [bob CLI](https://github.com/hashicorp/bob) that automatically bumps the version to a new patch. This automation has been introduced to crt-workflows-common as a [new workflow](https://github.com/hashicorp/crt-workflows-common/pull/233) (with the new bob command)  and handles version bumping at the end of the release pipeline (after being released to production); for example, `1.0.0→1.0.1` and `1.0.0-dev→1.0.0`. Bumping the minor version (ie `1.0.x→1.1.0`) is only supported manually via `bob update version -bump minor`, but not supported in CRT (this work is upcoming).  This is made possible by adding the new event “bump-version-patch” in the ci.hcl file in this PR. 

What this small action does: 

-  Allows for the static version string from the `.release/VERSION` file to be read by the new CRT workflow and automagically be bumped to the next version (whether it be a minor, or patch, or major version bump). 
- Outputs an error if there’s no `VERSION` file in the `.release/ dir`
- Outputs an error if there’s no version string in the VERSION file
- Is able to parse `product_version` if it is `1.3.0-alpha1` as `1.3.0` (example: when `product_version = 1.3.0-alpha1`, `base_version = 1.3.0`)
- Is able to parse prerelease product versions such as `alpha1` (example `prerelease_product_version = alpha1`) in the statement above.  

In other words, the action grabs the version from the version file at the beginning of the build, then passes those versions (along with metadata, where necessary) to the new version bump automation CRT workflow at the end of the release pipeline to bump the product’s version. One thing to note is that the `version.go` and `base_version.go` files are used for local development and will be now overwritten by the ld-flags with the actual version that are being called in the go builds. 

This PR also introduces [actions-go-build](https://github.com/hashicorp/actions-go-build), an action that dynamically takes care of building and packaging go binaries. Through this action, you can define the go build, assert that it’s reproducible (optional*) and then use the artifacts in the workflow. (Once the build action run is finished, you can check out it the builds are reproducible by scrolling down on the summary page, [example](https://github.com/hashicorp/packer/actions/runs/3587803756#summary-9808859635)) 

*NOTE:  if [assert is enabled](https://github.com/hashicorp/actions-go-build#reproducibility-assertions), this could potentially create some disruptions in workflows as the action tries to verify two builds (primary build and ui builds) are the same (size, shasums) as ui builds might have non-deterministic compilation.

Passing tests ✔️  Check out these examples:

- https://github.com/hashicorp/consul-enterprise/actions/runs/3623272718

This PR also introduces a bump to the version tags for `actions/checkout`, `actions/setup-go`, and `actions/upload-artifact` (setup-go version tag update was required by the introduction of actions-go-build)-- however, if the Packer team feels like these should be isolated in their own PRs for easier/broader visibility, super happy to raise separate PRs! 

Finally, as a part of this effort to onboard new CRT tooling, we have implemented the new `prepare` workflow in the `ci.hcl` config file. This workflow simply wraps all workflows in the pre-stable channel in one workflow, to make your ci.hcl config file more readable for your team. The `prepare` runs `upload-dev`, `security-scan`, `notarize`, `sign`, `verify`, `promote-dev-docker`, and `fossa-scan` (depending on your team’s current needs) under the hood - no need to worry about these workflows anymore!

# TODO Before Merging
 - [ ] Update version/VERSION to latest version
 - [ ] Remove `set-product-version` branch from `build.yml` workflow dispatch list
 - [ ] Remove `set-product-version` branch from `ci.hcl` allowlist

---

<details>
<summary> Overview of commits </summary>

  - a98fa8de9a07ad14f235a9a98ce9da65c43ecec1  - 8152ecf11d3efb73c0d58bd02ae806c4b01e339f  - e9f33a1f4a2719577f459863c94c4801605a4410  - be97ac97f6fe9ea6e8aa2e71cd9e8b48b2475633  - 83dbb2342cd565a833002e7966a3d32df1b70bac  - 8c98f5fde96dba8c1c7d2c6ebb388abeaed9e9e6  - 78ee14f6294c8d2da8879c6b77b5d3c7553a8d15  - 5e23cd7c7ff728f6d9486d01a12dc03293d4e77b  - 607b309db963a0b1f8ee8c5051ac944b3ef631fc  - 1bb4b45b3ff7a3ef9ebf35c5dcf4cffa2c9536c2  - b3107771dffd0f5c25a5f18a584c2d96e3a0df9a  - 5aaf6e4d8915ee71a518b859da19c8a48e989cd4  - 0546c6ec118c106fbdd505d2de42d7edb706b5be  - 2ff351aa47f29ed36e321052c0f1e244e385f136  - 4c097b85e38721028e0c619295f5c2025952d35b  - 9bfc23d1785cef83aab9695dae641a8e6b41ce77  - 473f2819d109ad5abf3e6d1c73f22d796eea2d62  - 6ef395d1ae1a88acec7068b19b9acc4077b131f3  - c977e4c74d57104bad90e3b851d174d1acb48fb3  - 5caad41d6288e5ec50cb761c098edfb086e4d185  - c3b96fbf957922e5426162187479321addb82994  - e9d0646778ac7571458aefee008592a5365fb8c2  - 71b7e46188e33b0c09576360611ff6d2e01d4f83  - 6e5129cb934d37b3f72513d75d14d5088a37b579  - fc4a05f148f51d00321d99c8d8c8e231c7bb8c19  - e14b64558779b2d879ad8b9b7b4928bbb1c52b46  - 046afb2b80e5b4cfaabd63d9d12d8a9e6cfcee3f  - 50984a9a67d24e76b21b1db82ff7bde378416a4f  - 92e34a48e1539cfca3b807a35b5005aa625edbbe  - c55176614756cb010f29399a1ffc2db60813af2e  - bbd1cb07b530d830712ac88e73a94df250421fad  - 76a14dcb575e8b252d43f14f90bffaad21488679 

</details>

### CODE REVIEW

**Review by github-team-consul-core-pr-approver — APPROVED** [review]
Auto approved Consul Bot automated PR
